### DISCUSSION TITLE
WASM Support

### DISCUSSION THREAD

**Discussion by John0x** [dataplane-app/dataplane#315] [Ideas & feature requests]
Hey, first of all, this looks extremely promising. I've been looking for a modern airflow alternative for quite some time. Keep up the great work 👍

Have you thought about supporting languages apart from python? Web Assembly would be perfect for something like this. The reliance on python has been one of my biggest pain points with most data pipelines.

**Comment by saul-data** [comment]
Hi there @John0x, glad you find it promising and thanks for the feedback. I haven't worked with WASM before but my understanding is that its more web browser side. Are you referring to the web browser code editor and running code on the browser? On the server side, essentially any language that can run in Linux can be bundled into the workers for example Rust and called using the Bash processor - this can call any command in bash/sh. If you send me a hello world / basic example of what you are trying to achieve and give a bit of explanation, I can see how it might be built into a worker docker image.

**Comment by John0x** [comment]
Ah, sorry, didn't see your response. I meant WASI on the server side with proper integration for the code editor. It would be pretty awesome to have the same experience that you currently have for Python, for Rust or other languages.

**Reply by saul-data** [reply]
So the key here is about other coding language support?

Let's take Rust for example which requires being compiled. Would that be edited in code editor then compiled to WASI/M then run?

I would need to understand that flow.

**Reply by saul-data** [reply]
Dataplane uses Monaco which is essentially vs code so the writing of the code and saving the files down is not an issue. 

The issue is how the code gets compiled and run. 

We are using Kubernetes to run Dataplane so another question is how this fits into those kind of deployment types.

Just trying to get a sense of how this will work end to end.

If there is a quick prototyping using Monaco editor would be greatly helpful.

**Comment by saul-data** [comment]
@John0x @gedw99 Would the below work? I think the easiest way of using different coding languages with Dataplane is simply to copy the compiled binary into workers through your CI/CD cycle.

```Dockerfile
FROM dataplane/dataplane-worker-python:latest

RUN mkdir -p /customcode && && chown -R appuser:appuser /customcode
COPY RustBinary /customcode/RustBinary
```

Then in Dataplane use a bash processor, change the command to run the binary. 

<img width="866" alt="image" src="https://user-images.githubusercontent.com/63714857/200399929-ec3ebaad-9401-480b-9d50-6521a61020df.png">

**Reply by John0x** [reply]
> @John0x @gedw99 Would the below work? I think the easiest way of using different coding languages with Dataplane is simply to copy the compiled binary into workers through your CI/CD cycle.
> 
> ```dockerfile
> FROM dataplane/dataplane-worker-python:latest
> 
> RUN mkdir -p /customcode && && chown -R appuser:appuser /customcode
> COPY RustBinary /customcode/RustBinary
> ```
> 
> Then in Dataplane use a bash processor, change the command to run the binary.
> 
> <img alt="image" width="866" src="https://user-images.githubusercontent.com/63714857/200399929-ec3ebaad-9401-480b-9d50-6521a61020df.png">

Well, it's definitely better than not being able to use rust at all, but I think it would be a killer feature to have the full experience for other languages as well(at least a few popular ones).
It's not an easy task, that's for sure, but it would be a huge selling point, in my opinion.

https://github.com/infinyon/fluvio
Allows to use Rust with WASM as sandboxed workers.

For the whole build pipeline stuff, you could probably take some inspiration from CI platforms such as Railway. They use nixpacks to support a multitude of languages: https://docs.railway.app/deploy/builds

**Reply by saul-data** [reply]
@John0x thanks that's helpful. Yes I agree this could be very powerful, I now understand where you are both coming from, this feature is now on my radar and I can look at your examples. We have some priority items that we are busy working on right now so this will take some time to architect and get in. I am hoping that workaround works for now to bridge the time/need gap. 😃

**Comment by saul-data** [comment]
Here is a good article on what would be required to get Rust to work. You can see the code needs to be compiled first. This is a good example because it uses Golang to run shell commands to compile the Rust code to WASM. This can be applied for WASI too. 

https://blog.boot.dev/rust/rust-in-the-browser/
